### PR TITLE
Add weak constraints to make rects closer to each other in size 

### PR DESCRIPTION
**Pull request by kdheepak**
This PR introduces a `SegmentSize` `enum` that will allow adding the following constraints optionally:

```rust

        for i in 0..elements.len() {
            for j in (i + 1)..elements.len() {
                solver.add_constraint(elements[j].size() | EQ(WEAK) | (elements[i].size()))?;
            }
        }

```

This makes chunks similar in size to each other if other constraints are satisfied.

This PR also changes the implementation of `Max` and `Min` by adding a `MEDIUM` equality constraint instead of a `WEAK` equality constraint. 

```diff
     Constraint::Max(m) => {
         solver.add_constraints(&[                           
             element.size() | LE(STRONG) | f64::from(m),     
-            element.size() | EQ(WEAK) | f64::from(m),
+            element.size() | EQ(MEDIUM) | f64::from(m),
         ])?;                                                
     }                                                       
     Constraint::Min(m) => {                                 
         solver.add_constraints(&[                           
             element.size() | GE(STRONG) | f64::from(m),     
-            element.size() | EQ(WEAK) | f64::from(m),
+            element.size() | EQ(MEDIUM) | f64::from(m),
         ])?;                                                
     }                                                       
```

It should result in the same layouts for us though, because there were no other `MEDIUM` constraints used.

### CODE REVIEW

File: src/layout.rs
Comment (joshka) [inline src/layout.rs:615]: What happens when this fails on a CI server somewhere because of some unrelated change (or due to something non-deterministic), meaning the failure is difficult to repro locally?

The first check failing here will prevent the 2nd to 4th checks executing, which contain the information you'd need to understand how to fix the test  A better test for this would show how those numbers change for various split widths:

```suggestion
    fn test_horizontal_split_equally_in_underspecified_case() {
        use Constraint::*;
        fn split_widths(constraints: &Vec<Constraint>, target_width: u16) -> Vec<u16> {
            let target = Rect::new(2, 2, target_width, 10);
            Layout::default()
                .direction(Direction::Horizontal)
                .constraints(constraints.as_ref())
                .split(target)
                .iter()
                .map(|r| r.width)
                .collect_vec()
        }

        let constraints = vec![Min(2), Min(2), Min(0)];
        assert_eq!(split_widths(&constraints, 0), vec![0, 0, 0]);
        assert_eq!(split_widths(&constraints, 1), vec![1, 0, 0]);
        assert_eq!(split_widths(&constraints, 2), vec![2, 0, 0]);
        assert_eq!(split_widths(&constraints, 3), vec![2, 1, 0]);
        assert_eq!(split_widths(&constraints, 4), vec![2, 2, 0]);
        assert_eq!(split_widths(&constraints, 5), vec![2, 2, 1]);
        assert_eq!(split_widths(&constraints, 6), vec![2, 2, 2]);
        assert_eq!(split_widths(&constraints, 7), vec![2, 2, 3]);
        assert_eq!(split_widths(&constraints, 8), vec![2, 2, 4]);
        assert_eq!(split_widths(&constraints, 9), vec![2, 2, 5]);
        assert_eq!(split_widths(&constraints, 10), vec![2, 2, 6]);
    }
```
Those are the numbers before the change to the split function is applied. Normally I much prefer to have tests be  linear and simpler than this one, but I think a good reason to find an exception is where the test communicates the relationship between the outputs of similar inputs that would be otherwise difficult to understand in isolation.

What this test doesn't communicate / test is where the edge cases on the new code take effect.

Oddly, I can't actually get this to succeed with the new numbers as this now fails non-deterministically. Is the code intended to be deterministic?

E.g. here are the results of `cargo test` run several times:
```
thread 'layout::tests::test_horizontal_split_equally_in_underspecified_case' panicked at 'assertion failed: `(left == right)`
  left: `[1, 1, 1]`,
 right: `[1, 1, 0]`', src/layout.rs:596:9
```
```
thread 'layout::tests::test_horizontal_split_equally_in_underspecified_case' panicked at 'assertion failed: `(left == right)`
  left: `[1, 0, 0]`,
 right: `[1, 1, 0]`', src/layout.rs:596:9
```
```
thread 'layout::tests::test_horizontal_split_equally_in_underspecified_case' panicked at 'assertion failed: `(left == right)`
  left: `[0, 0, 1]`,
 right: `[0, 0, 0]`', src/layout.rs:595:9
```
```
thread 'layout::tests::test_horizontal_split_equally_in_underspecified_case' panicked at 'assertion failed: `(left == right)`
  left: `[2, 2, 0]`,
 right: `[1, 1, 0]`', src/layout.rs:598:9
```

File: src/layout.rs
Comment (kdheepak) [reply src/layout.rs:321]: I actually like keeping all the constraints in the same function. What do you think about me adding comments instead?

File: src/layout.rs
Comment (kdheepak) [reply src/layout.rs:615]: This is a can of worms if it is non-deterministic. Maybe we should only have tests that are deterministic?

Comment (joshka) [reply src/layout.rs:615]: Yes, we should only have deterministic tests, but so too behavior.

Ideally we need to find a way to push the constraints to choose more deterministic values for constraints - perhaps by being more granular with the weighting of the importance of each constraint? I think this probably becomes complex fast though, but the worms are there already I think.

File: src/layout.rs
Comment (joshka) [reply src/layout.rs:321]: My general rationale for adding a function instead of just commented sections is that everything you need to understand in a particular block is limited to the scope of the function. A commented section could be affected by any change above, and any change within can effect any code below, which means you have to understand the full function to maintain it. In any particular part of the lib, that's not a big deal, but when large functions dominate, every maintenance task tends to become larger as the time spent understanding is magnified.

Comment (kdheepak) [reply src/layout.rs:321]: Fair. In that case I think I’d like to move all the different constraints into individual functions too then.

File: src/layout.rs
Comment (hasezoey) [inline src/layout.rs:44]: i dont know if this change was automatic, but removing the double space at the end will make the lines as if it were a single line, either re-add the double space, or add a empty line inbetween

Comment (joshka) [reply src/layout.rs:44]: Let's use an empty line rather than invisible characters that affect things. The rust-doc convention is that the first line of a doc comment is a single sentence paragraph, not multiple lines with a line break.

**Review by hasezoey — CHANGES_REQUESTED** [review]
i know this a bit unrelated, but this change causes some side-effects, please test against #404 with this change applied, see https://github.com/ratatui-org/ratatui/issues/370#issuecomment-1678722149 for more context

**Comment by joshka** [comment]
> i know this a bit unrelated, but this change causes some side-effects, please test against #404 with this change applied, see [#370 (comment)](https://github.com/ratatui-org/ratatui/issues/370#issuecomment-1678722149) for more context

The main blocker for this PR right now is that it results in non-deterministic splits more often, which is makes it difficult to reason about how space will be allocated for users.

**Comment by kdheepak** [comment]
I did some testing with this and wrote a integer optimization program (in Julia) to simulate "ideal" behavior to compare. I think we are never going to be able to guarantee that there won't be gaps with cassowary. 

For example, take a constraints `Percent(33), Percent(33), Percent(33)` in a `Horizontal` layout of `area` whose width is 100. The "ideal" result is going to be `33.3333`, `33.3333` and `33.3333` for each `Rect`, and rounded you are going to get `33`, `33`, `33` which will leave a gap of `1`.

I implemented [this using integer variables in Julia](https://gist.github.com/kdheepak/afc1ed987c19d67d738661559c2cc6b5) and I get a solution that we would expect, i.e. `33`, `34` and `33`.

```julia
    area = Rect(0, 0, 100, 1)
    constraints = [Percent(33), Percent(33), Percent(33)]
    r1, r2, r3 = split(Horizontal(constraints), area; kwargs...)
    @test r1.width == 33
    @test r2.width == 34
    @test r3.width == 33
```

image

Unfortunately, this behavior is not guaranteed with cassowary because cassowary uses floating point values.

**Comment by kdheepak** [comment]
Adding the weak constraints to make the widths (or heights) similar did have the intended effect (imo). Take this test for example:

```julia
    area = Rect(0, 0, 100, 1)
    constraints = [Percent(50), HardMin(10), Percent(50)]
    r1, r2, r3 = split(Horizontal(constraints), area; kwargs...)
    @test r1.width == 45
    @test r2.width == 10
    @test r3.width == 45
```

Without the weak constraints, I was getting results like this:

```julia
    area = Rect(0, 0, 100, 1)
    constraints = [Percent(50), HardMin(10), Percent(50)]
    r1, r2, r3 = split(Horizontal(constraints), area; kwargs...)
    @test r1.width == 50
    @test r2.width == 10
    @test r3.width == 40
```

**Comment by joshka** [comment]
>I think we are never going to be able to guarantee that there won't be gaps with cassowary.

The approach that makes this round without gaps is to use the left/right position rather than the widths to work out how to round.
Thus instead of thinking of [33.33, 33.33, 33.33] => [33, 33, 33], think instead of [0.00, 33.33, 66.66, 100.00] => [0, 33, 67, 100]

This works pretty nicely as far as I can tell and is my intended change as soon as the refactor in #405 lands.

**Comment by kdheepak** [comment]
Interesting idea! I think there are some corner cases there as well (I'll have to test to come up with examples), but maybe fewer issues than what we have at the moment, so it should be good!

**Comment by joshka** [comment]
> Interesting idea! I think there are some corner cases there as well (I'll have to test to come up with examples), but maybe fewer issues than what we have at the moment, so it should be good!

Perhaps something that evaluates to (positions): `[0.0, 0.51, 1.49, 2]` => `[0, 1, 1, 2]`. The widths would be `[0.51, 0.98, 0.51]`, but become `[1, 0, 1]` after rounding.

File: src/layout.rs
Comment (kdheepak) [reply src/layout.rs]: With the recent refactors, I'm assuming it is okay to have the equations in the same function.

File: src/layout.rs
Comment (kdheepak) [reply src/layout.rs]: I ran it multiple times locally and it doesn't appear to be non-deterministic for me, although I'm not sure why since I thought the constraints were essentially the same as before.

File: src/layout.rs
Comment (kdheepak) [reply src/layout.rs]: This has been resolved in another PR.

**Review by hasezoey — COMMENTED** [review]
code looks good to me, though it would probably be great to also have some tests for `Max` and `Length` to ensure they are not somehow heigher / lower than what they should be

**Comment by kdheepak** [comment]
I added tests for `Max` and `Length` too. Open to better suggestions for the test for `Max`.

File: src/layout.rs
Comment (kdheepak) [inline src/layout.rs:758]: I don't understand why this test is behaving this way. This seems like a bug to me.

Comment (kdheepak) [reply src/layout.rs]: Fixed in https://github.com/ratatui-org/ratatui/pull/395/commits/0b41dac3944df4c052ab872385d6f2bf926d3dd1

This commit changes the implementation of `Max` and `Min` by adding a `MEDIUM` constraint instead of a `WEAK` constraint. It should result in the same layouts for us though, because there were no other `MEDIUM` constraints used.

Comment (joshka) [reply src/layout.rs]: Good fix!

**Review by joshka — CHANGES_REQUESTED** [review]

File: src/layout.rs
Comment (kdheepak) [reply src/layout.rs]: Addressed in https://github.com/ratatui-org/ratatui/pull/395/commits/17790b7fbd2a3f9252c0f8b6b63f4560fd904920. Let me know if you have better suggestions for naming.

File: src/layout.rs
Comment (kdheepak) [reply src/layout.rs]: Done in https://github.com/ratatui-org/ratatui/pull/395/commits/17790b7fbd2a3f9252c0f8b6b63f4560fd904920

File: src/layout.rs
Comment (kdheepak) [reply src/layout.rs:640]: Good points, I meant to do simplifications to the tests myself but didn't get to it in time.

File: src/layout.rs
Comment (kdheepak) [reply src/layout.rs]: I changed it to simplify the tests. Hopefully this change in https://github.com/ratatui-org/ratatui/pull/395/commits/17790b7fbd2a3f9252c0f8b6b63f4560fd904920 will address this comment.

File: src/layout.rs
Comment (kdheepak) [reply src/layout.rs]: I didn't choose the letters approach, which I like, because we are testing a width of `100` here and it was a very long string.

File: src/layout.rs
Comment (kdheepak) [inline src/layout.rs:296]: I would have liked to do this:

```
            chunk_size_preference: Default::default(),
```

But I can't call `non-const` in `const` functions.

Comment (joshka) [reply src/layout.rs:296]: Yeah - it's really annoying that trait methods like `Default::default()` can't be const :/ There's an RFC to fix it though - so someday...

File: src/layout.rs
Comment (joshka) [reply src/layout.rs]: What about something like AllocateLeftoverSpace (and have this incorporate the expand to fill option, which is pub(crate) so can be removed without breaking anything)
```rust
enum AllocateLeftoverSpace {
  None,
  Last,
  Equal,
}
```
(I'm never quite sure whether to add an explicit `None` vs using Option though)

File: src/layout.rs
Comment (joshka) [inline src/layout.rs:643]: nit: Can this either pass the ChunkSizePreference directly (add `use ChunkSizePreference::*` to make it concise) or create two methods. Generally bool params mean that people have to scroll up to work out the meaning.

**Review by joshka — CHANGES_REQUESTED** [review]
This looks great now. I think this is the right way to make layout more responsive.

File: src/layout.rs
Comment (kdheepak) [reply src/layout.rs:237]: I like the idea of using a single `enum` for capturing the behavior of "dumping excess into the last chunk" _and_ "prefer making chunks equal in size".

I think one issue is that "prefer making chunks equal in size" can be done even when there's no "leftover space", whereas "dumping excess into the last chunk" only happens when there's "leftover space".

1. Should we combine these concepts that are slightly different? I'm leaning toward yes.
2. If so, what should the name of the `enum` be? I think `AllocateLeftoverSpace` underplays what happens when you choose `Equal`.

Comment (joshka) [reply src/layout.rs:237]: Asking for ideas in https://discord.com/channels/1070692720437383208/1072879985762762812/1142281786974544065

Comment (kdheepak) [reply src/layout.rs:237]: Maybe this?

```rust
#[derive(Debug, Default, Clone, Eq, PartialEq, Hash)]
pub enum AllocationPreference {
    #[default]
    None,
    LastGathersRemainder,
    AllEqual,
}
```

Also, do we want to allow `expand_fill: true` + `AllEqual`? Using the enum doesn't allow for that unless we add a new enum member, and in that case I don't know what to name it.

File: src/layout.rs
Comment (kdheepak) [reply src/layout.rs]: Fixed in https://github.com/ratatui-org/ratatui/pull/395/commits/819ed54821b6c4abe2c5e8d826e1da34cf5e0d52.

File: Cargo.toml
Comment (kdheepak) [reply Cargo.toml]: Fixed in https://github.com/ratatui-org/ratatui/pull/395/commits/819ed54821b6c4abe2c5e8d826e1da34cf5e0d52.

File: src/layout.rs
Comment (kdheepak) [reply src/layout.rs]: I've settled for this for now:

```rust
#[derive(Debug, Default, Clone, Eq, PartialEq, Hash)]
pub enum SegmentSizePreference {
    EqualDistribution,
    #[default]
    LastTakesRemainder,
    None,
}
```

`AllEqual` also activates `LastGathersRemainder`.

**Review by joshka — APPROVED** [review]
Approved, but let's hold off a minute to workout the right naming for this.

We could alternatively keep this `pub(crate)` but expose it as `pub` within an "experimental" feature flag to help make it obvious that this might change.

File: src/layout.rs
Comment (hasezoey) [reply src/layout.rs]: > (I'm never quite sure whether to add an explicit None vs using Option though)

personally, if it is a enum similar to this, then i (as a user) would prefer to just have to define `AllocateLeftoverSpace::None` and `AllocateLeftoverSpace::Equal` instead of `(Option::)None` and `Some(AllocateLeftoverSpace::Equal)`

also as for a techincal aspect, this *could* save one WORD(bytes) of space, because of the option identification

**Comment by kdheepak** [comment]
> i know this a bit unrelated, but this change causes some side-effects, please test against #404 with this change applied, see [#370 (comment)](https://github.com/ratatui-org/ratatui/issues/370#issuecomment-1678722149) for more context

I had to fix some tests after #404 was merged and after I rebased on top of it.

File: src/widgets/table.rs
Comment (hasezoey) [inline src/widgets/table.rs:761]: something is not right here, they should be the same width (there is more than available space)

Comment (kdheepak) [reply src/widgets/table.rs:761]: I think this test is not deterministic? I stashed all my changes and ran this test on `main`; and I'm always getting a failure.

image

Comment (kdheepak) [reply src/widgets/table.rs:761]: When I run on `main` with these two changes:

```rust
            dbg!(changes.get(&element.start));
            dbg!(changes.get(&element.end));
```

```
$ cargo test -- widgets::table::tests::table_column_widths::ratio_constraint --nocapture
running 1 test
[src/layout.rs:568] changes.get(&element.start) = None
[src/layout.rs:569] changes.get(&element.end) = Some(
    0.0,
)
[src/layout.rs:568] changes.get(&element.start) = Some(
    0.0,
)
[src/layout.rs:569] changes.get(&element.end) = Some(
    6.666666666666666,
)
[src/layout.rs:568] changes.get(&element.start) = Some(
    6.666666666666666,
)
[src/layout.rs:569] changes.get(&element.end) = Some(
    7.666666666666666,
)
[src/layout.rs:568] changes.get(&element.start) = Some(
    7.666666666666666,
)
[src/layout.rs:569] changes.get(&element.end) = Some(
    14.333333333333332,
)
thread 'widgets::table::tests::table_column_widths::ratio_constraint' panicked at 'assertion failed: `(left == right)`
  left: `[(0, 7), (8, 6)]`,
 right: `[(0, 6), (7, 6)]`', src/widgets/table.rs:752:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test widgets::table::tests::table_column_widths::ratio_constraint ... FAILED

failures:

failures:
    widgets::table::tests::table_column_widths::ratio_constraint
```

I think in #404 there was no rounding, so your tests were passing then. But #408 introduced rounding.

There's something I don't understand though. Here's the git log history:

image

You can see that #408 was before #404. After merging #404 to `main`, the tests were indeed failing: https://github.com/ratatui-org/ratatui/actions/runs/5911487064/job/16034036047.

But my understanding was that GitHub's merge queue would run the tests again on the latest HEAD before merging; which would have caught this issue before merging #404. Does anyone know why that did not happen? Is there a feature that we need to enable on the merge queue feature on GitHub?

Comment (hasezoey) [reply src/widgets/table.rs:761]: > You can see that https://github.com/ratatui-org/ratatui/pull/408 was before https://github.com/ratatui-org/ratatui/pull/404. And after merging https://github.com/ratatui-org/ratatui/pull/404 to main, the tests were indeed failing on CI: https://github.com/ratatui-org/ratatui/actions/runs/5911487064/job/16034036047.

after some quick tests, i came to the same conclusion

> But my understanding was that GitHub's merge queue would run the tests again on the latest HEAD before merging; which would have caught this issue before merging https://github.com/ratatui-org/ratatui/pull/404. Does anyone know why that did not happen? Is there a feature that we need to enable on the merge queue feature on GitHub?

i was also under that assumption, but maybe the branch-protection on `main` does not require all tests to pass? i came to this conclusion because of the following doc snippit

> Once a pull request has passed all required branch protection checks, a user with write access to the repository can add the pull request to the queue. The merge queue will ensure the pull request's changes pass all required status checks when applied to the latest version of the target branch and any pull requests already in the queue.

another possibility is that maybe because [of these lines](https://github.com/ratatui-org/ratatui/blob/b6b2da5eb761ac5894cc7a2ee67f422312b63cfc/.github/workflows/ci.yml#L34-L38) the *pull request head* was checked out *instead of the merged head*, but i am not sure, because i dont know if [this run](https://github.com/ratatui-org/ratatui/actions/runs/5911487064) is the run that tested the merge-ability, or the test after the merge

Comment (hasezoey) [reply src/widgets/table.rs:761]: for now i have created a meta-issue to track this instead of this PR https://github.com/ratatui-org/ratatui/issues/418

Comment (kdheepak) [reply src/widgets/table.rs:761]: Great, I'll resolve this conversation.

**Comment by kdheepak** [comment]
I renamed it to `SegmentSize` and made it `pub(crate)` so that we can finalize the name later with more feedback.

File: src/layout.rs
Comment (kdheepak) [reply src/layout.rs]: I renamed it to `SegmentSize` and made it `pub(crate)` so that we can finalize the name later with more feedback.

File: src/widgets/table.rs
Comment (joshka) [reply src/widgets/table.rs]: I fixed the table tests that were broken in https://github.com/ratatui-org/ratatui/pull/419

**Comment by joshka** [comment]
I pushed a rebase on the fix in #419 - happy for this to be approved and merged now @kdheepak. We can work out the public name for the API at some point later. I lean towards using `Allocation` as the name and communicating that this is a preference not a requirement in the enum variant itself, and also using an imperative tone that says what the allocation will do (e.g. `PreferEvenWidths`, `FillLastArea`, etc.)

**Review by joshka — APPROVED** [review]

**Comment by kdheepak** [comment]
I did the rebase locally as well, you are faster than me :) I'll discard my changes.

Yes, let's merge and figure out the right names for it later. I'm not tied to any names here.

**Review by joshka — APPROVED** [review]

**Review by sophacles — APPROVED** [review]

**Pull request merged**
